derive/src/lib.rs

```diff
@@ -108,11 +108,47 @@ pub fn decode_derive(input: TokenStream) -> TokenStream {
 	generated.into()
 }
 
-fn add_trait_bounds(mut generics: Generics, bounds: syn::TypeParamBound) -> Generics {
-	for param in &mut generics.params {
-		if let GenericParam::Type(ref mut type_param) = *param {
-			type_param.bounds.push(bounds.clone());
+fn add_trait_bounds(generics: &mut Generics, data: &syn::Data, bound: syn::Path) {
+	let types = collect_types(&data);
+	if !generics.params.is_empty() && !types.is_empty() {
+
+		if generics.where_clause.is_none() {
+			generics.where_clause = Some(parse_quote!(where));
 		}
+
+		let where_clause = &mut generics.where_clause.as_mut().unwrap();
+
+		for ty in types {
+			where_clause.predicates.push(parse_quote!(#ty : #bound));
+		}
+	}
+}
+
+fn collect_types(data: &syn::Data) -> Vec<syn::Type> {
+	use syn::*;
+
+	match *data {
+		Data::Struct(ref data) => match &data.fields {
+			| Fields::Named(FieldsNamed { named: fields , .. })
+			| Fields::Unnamed(FieldsUnnamed { unnamed: fields, .. }) => {
+				fields.iter().map(|f| f.ty.clone()).collect()
+			},
+
+			Fields::Unit => { vec![] },
+		},
+
+		Data::Enum(ref data) => data.variants.iter().flat_map(|variant| {
+			match &variant.fields {
+				| Fields::Named(FieldsNamed { named: fields , .. })
+				| Fields::Unnamed(FieldsUnnamed { unnamed: fields, .. }) => {
+					fields.iter().map(|f| f.ty.clone()).collect()
+				},
+
+				Fields::Unit => { vec![] },
+			}
+		})
+		.collect(),
+
+		Data::Union(_) => panic!("Union types are not supported."),
```

Comment:
Please no `panic`, make the function return a Result and at the point where you generate code, use https://docs.rs/syn/0.15.26/syn/parse/struct.Error.html#method.to_compile_error
This will show a nice compile error.

Reply:
Thanks for the suggestion. However, that's how it was implemented before. Also, code in `encode` and `decode` modules panics too.

Reply:
Yeah, but from an usage perspective it is not nice. If we touch this code, we can improve it :)

Reply:
Ok, will change that code as well.